body.tex

```diff
@@ -1,166 +1,466 @@
 
-\subsubsection{Telescope and Site}
+Adding the support for AWS into the Butler exercised a significant fraction of the Generation 3 Data Butler. During the process many faults and unpredictable behaviors were discovered and solved. Many problems touched, and continue to exercise, the general Generation 3 Data Butler implementation, as well as assumptions made during their implementation. Recounting the wide list of major improvements to the codebase, hopefully, reveals how productive this exercise has been in helping generalizing and strengthening the whole Generation 3 Data Butler codebase.
 
-\begin{itemize}
-   \item Telescope and site overview and status in 2014:  \cite{2014SPIE.9145E..1AG}
-\end{itemize}
-%------------------------------------------------------------------------------
+\subsection{PostgreSQL performance}
 
-\subsubsection{System Engineering}
+The following described performance issue applies strictly to w\_2019\_38 builds and earlier. The QuantumGraph is created by issuing a very large SQL statement that, effectively, creates a cartesian product between many, of the 15-ish or so, tables in the registry. The results of the large query are then parsed in Python and a slew of many different, small, follow-up queries are issued. This diverse workflow presents a difficult challenge in DB optimization. The Generation 3 Middleware team reports that on average it takes 0.5 to 1.5h to create QuantumGraphs for tract-size DRP workflows with Oracle on NCSA infrastructure. In our tests we were unable to create even the simplest QuantumGraphs, even after doubling of the RDS instance resources, and were forced to abort multiple times after 30+h of execution. The lack of performance did not seem related to hardware limitations, since none of the performance metrics availible showed heavy hardware loads nor didthe problem go away even after significant increase of RDS resources.
 
-\begin{itemize}
-   \item LSST systems engineering: \cite{2014SPIE.9150E..0MC}
-   \item System verification and validation: \cite{2014SPIE.9150E..0NS}
-\end{itemize}
-%
+PostgreSQL, and Oracle, regularly collect statistics on database objects used in queries. These statistics are then used to generate execution plans. These execution plans are then oftentimes cached. This in general results in performance gains. But because Butler generates all SQL dynamically there are no guarantees that a cached, or even pre-prepared execution plan can be executed. In PostgreSQL many of these optimizations then fail.
```

Comment:
I don't think it was demonstrated that cached execution plans had anything to do with the problem.  I would remove this paragraph.

Reply:
Selecting from a view is as fast or as slow as the underlying SQL subquery statement - at least it should be in ideal conditions. To guarantee the best performance recommendation is to include conditions in the view's definition. The optimizer tries to create the optimal execution plan by pushing the predicates on a view into the view before executing the statement but for any/most RDBMS this can sometimes fail for complex queries where conditions are several layers away from the view. 

Defining functions, preparing and caching uniquely optimized execution plans for these large queries helps avoid situations where optimizer fails. We don't use prepared or cached statements - but perhaps they would help. The bonus point is that the way the SQL statements are generated sometimes works against PostgreSQL's internals and similar situations, where we expect the RDBMS to just work without providing it with sufficient information, will continue to occur. This is a consequence of the current registry design.